dartsim/src/JointFeatures_TEST.cc

```diff
@@ -883,6 +883,142 @@ TEST_F(JointFeaturesFixture, JointAttachDetach)
+  const std::string bodyName{"link"};
+
+  auto model1 = world->GetModel(modelName1);
+  auto model2 = world->GetModel(modelName2);
+  auto model3 = world->GetModel(modelName3);
+
+  auto model1Body = model1->GetLink(bodyName);
+  auto model2Body = model2->GetLink(bodyName);
+  auto model3Body = model3->GetLink(bodyName);
+
+  const dart::dynamics::SkeletonPtr skeleton1 =
+      dartWorld->getSkeleton(modelName1);
+  const dart::dynamics::SkeletonPtr skeleton2 =
+      dartWorld->getSkeleton(modelName2);
+  const dart::dynamics::SkeletonPtr skeleton3 =
+      dartWorld->getSkeleton(modelName3);
+  ASSERT_NE(nullptr, skeleton1);
+  ASSERT_NE(nullptr, skeleton2);
+  ASSERT_NE(nullptr, skeleton3);
+
+  auto *dartBody1 = skeleton1->getBodyNode(bodyName);
+  auto *dartBody2 = skeleton2->getBodyNode(bodyName);
+  auto *dartBody3 = skeleton3->getBodyNode(bodyName);
+
+  ASSERT_NE(nullptr, dartBody1);
+  ASSERT_NE(nullptr, dartBody2);
+  ASSERT_NE(nullptr, dartBody3);
+
+  const math::Pose3d initialModel1Pose(0, -0.2, 0.45, 0, 0, 0);
+  const math::Pose3d initialModel2Pose(0, 0.2, 0.45, 0, 0, 0);
+  const math::Pose3d initialModel3Pose(0, 0.6, 0.45, 0, 0, 0);
+
+  EXPECT_EQ(initialModel1Pose,
+            math::eigen3::convert(dartBody1->getWorldTransform()));
+  EXPECT_EQ(initialModel2Pose,
+            math::eigen3::convert(dartBody2->getWorldTransform()));
+  EXPECT_EQ(initialModel3Pose,
+            math::eigen3::convert(dartBody3->getWorldTransform()));
+
+  physics::ForwardStep::Output output;
+  physics::ForwardStep::State state;
+  physics::ForwardStep::Input input;
+
+  // Create the first joint. This should be a normal fixed joint.
+  const auto poseParent1 = dartBody1->getTransform();
+  const auto poseChild1 = dartBody2->getTransform();
+  auto poseParentChild1 = poseParent1.inverse() * poseChild1;
+  auto fixedJoint1 = model2Body->AttachFixedJoint(model1Body);
+  fixedJoint1->SetTransformFromParent(poseParentChild1);
+
+  EXPECT_EQ(initialModel1Pose,
+            math::eigen3::convert(dartBody1->getWorldTransform()));
+  EXPECT_EQ(initialModel2Pose,
+            math::eigen3::convert(dartBody2->getWorldTransform()));
+  EXPECT_EQ(initialModel3Pose,
+            math::eigen3::convert(dartBody3->getWorldTransform()));
+
+  // Create the second joint. This should be a WeldJoint constraint
+  const auto poseParent2 = dartBody3->getTransform();
+  const auto poseChild2 = dartBody2->getTransform();
+  auto poseParentChild2 = poseParent2.inverse() * poseChild2;
+  auto fixedJoint2 = model2Body->AttachFixedJoint(model3Body);
+  fixedJoint2->SetTransformFromParent(poseParentChild2);
+
+  EXPECT_EQ(initialModel1Pose,
+            math::eigen3::convert(dartBody1->getWorldTransform()));
+  EXPECT_EQ(initialModel2Pose,
+            math::eigen3::convert(dartBody2->getWorldTransform()));
+  EXPECT_EQ(initialModel3Pose,
+            math::eigen3::convert(dartBody3->getWorldTransform()));
+
+  const std::size_t numSteps = 100;
```

Comment:
Regarding the issue I mentioned, it can be recreated in this test if `numSteps` is bumped up to `100000`

Reply:
I've updated the test in https://github.com/gazebosim/gz-physics/pull/352/commits/fb11680f89dad22d10405d2db88908072e6907c5 to expect the falling velocity to be linearly increasing with time, which exposes the failure of on of the boxes to fall after detaching.

Reply:
the `DetachJoint` function is exiting [here](https://github.com/gazebosim/gz-physics/blob/fb11680f89dad22d10405d2db88908072e6907c5/dartsim/src/JointFeatures.cc#L380-L383) since the welded node is not in the `links` data structure

Reply:
potential fix in https://github.com/gazebosim/gz-physics/pull/352/commits/fe92bd5f1ac0b9895be2acd5adea0e3d6703cc9e